### PR TITLE
CD-244080 Add configurable options for: timeout, open_timeout, and se…

### PR DESCRIPTION
**Pull request by johnwu96822**
…conds_to_expire_token_before_expiry. 

Fix retry causing race conditions. Previously it deletes the token from cache before making the retry request. Now we don't delete token at all, and when the retry request gets a new token, it'll just overwrite the existing cached token.

- [ ] @johnny-lai 
- [ ] @supriya 
- [x] @abdulchaudhrycoupa

### CODE REVIEW

File: lib/sand/service.rb [hunk 1]
```diff
@@ -114,13 +117,26 @@ def check_token(token, options = {})
...
+        scopes: @scopes,
+        num_retry: options[:num_retry],
+        timeout: options.fetch(:timeout, timeout).to_i,
+        open_timeout: options.fetch(:open_timeout, open_timeout).to_i,
+        retrying: options[:retrying]
+      )
+      token_allowed_request(token, access_token, options)
+    end
 
+    def token_allowed_request(token, service_token, options = {})
```
Comment (johnwu96822) [inline lib/sand/service.rb:136]: I have wanted to split this method out from verify_token for a while. Easier to test.

File: lib/sand/service.rb [hunk 2]
```diff
@@ -114,13 +117,28 @@ def check_token(token, options = {})
...
+
+      resource = options.fetch(:resource, @resource).to_s.strip
+      raise ArgumentError.new("resource is required") if resource.empty?
+
+      service_token = self.token(
+        cache_key: SERVICE_CACHE_KEY,
+        scopes: @scopes,
+        num_retry: options[:num_retry],
+        timeout: options.fetch(:timeout, timeout).to_i,
+        open_timeout: options.fetch(:open_timeout, open_timeout).to_i,
```
Comment (abdulchaudhrycoupa) [inline lib/sand/service.rb:130]: @johnwu96822 can this take a float?

Comment (johnwu96822) [reply lib/sand/service.rb:130]: Good point

**Review by abdulchaudhrycoupa — APPROVED** [review]

**Review by abdulchaudhrycoupa — APPROVED** [review]

**Review by supriya — APPROVED** [review]

**Review by johnny-lai — APPROVED** [review]

**Comment by johnwu96822** [comment]
[lockit v4.9.3] merged via Lockit CLI on 2021-05-14 12:41:49.90245 -0700 PDT m=+4.882785522

**Pull request merged**